R/loo_moment_matching.R

```diff
@@ -0,0 +1,568 @@
+#' Moment matching for efficient approximate leave-one-out cross-validation (LOO)
+#'
+#' Moment matching algorithm for updating a loo object when Pareto k estimates
+#' are large.
+#'
+#'
+#'
+#'
+#' @export mmloo mmloo.default
+#' @param x A fitted model object.
+#' @param loo A loo object that is modified.
+#' @param post_draws A function the takes \code{x} as the first argument and
+#'   returns a matrix of posterior draws of the model parameters.
+#' @param log_lik A function that takes \code{x} and \code{i} and returns a
+#'   matrix (one column per chain) or a vector (all chains stacked) of
+#'   log-likeliood draws of the \code{i}th observation based on the
+#'   model \code{x}. If the draws are obtained using MCMC, the
+#'   matrix with MCMC chains separated is preferred.
+#' @param unconstrain_pars A function that takes arguments \code{x}, and
+#'   \code{pars} and returns posterior draws on the unconstrained space based on
+#'   the posterior draws on the constrained space passed via \code{pars}.
+#' @param log_prob_upars A function that takes arguments \code{x} and
+#'   \code{upars} and returns a matrix of log-posterior density values of the
+#'   unconstrained posterior draws passed via \code{upars}.
+#' @param log_lik_upars A function that takes arguments \code{x}, \code{upars},
+#'   and \code{i} and returns a vector of log-likelihood draws of the \code{i}th
+#'   observation based on the unconstrained posterior draws passed via
+#'   \code{upars}.
+#' @param max_iters Maximum number of moment matching iterations. Usually this
+#' does not need to be modified. If the maximum number of iterations is reached,
+#' there will be a warning, and increasing \code{max_iters} may improve accuracy.
+#' @param k_thres Threshold value for Pareto k values above which the moment
+#'   matching algorithm is used. The default value is 0.5.
+#' @param split Logical; Indicate whether to do the split transformation or not
+#'   at the end of moment matching for each LOO fold.
+#' @param cov Logical; Indicate whether to match the covariance matrix of the
+#' samples or not.
+#'   If \code{FALSE}, only the mean and marginal variances are matched.
+#' @template cores
+#' @param ... Further arguments passed to the custom functions documented above.
+#'
+#' @return The `mmloo()` methods return an updated \code{loo} object.
+#' The structure of the updated \code{loo} object is similar, but the
+#' method also stores the original Pareto k diagnostic values in
+#' the diagnostics field.
+#'
+#' @details The `mmloo()` function is an S3 generic and we provide a default
```

Comment:
We should indicate here that all of these functions should contain a `...` argument even if not actually used as otherwise they will break if any kind of additional argument is passed to one of these functions.

Reply:
Done!